### ISSUE TITLE
Workaround to deal with invalid headers?

### ISSUE THREAD

**Issue by keflavich** [astropy/astropy#5396]
CASA 4.7 was just released and it appears to produce invalid FITS headers, at least sometimes:

```
ValueError: FITS header values must contain standard printable ASCII characters; "'CASA 4.7.0-REL (r38335)\n'" contains characters not representable in ASCII or non-printable characters.
```

There is already a bug report upstream to fix this, but is there a workaround for the files that are created by this rogue version?  These files will fail whenever the header is accessed or written.

Note that `fh.verify('fix')` results in the same traceback.

(edit: tested on astropy 1.1.2)

**Comment by MSeifert04** [comment]
It would be interesting to know which unprintable characters. Could you share one "corrupted" file?

**Comment by keflavich** [comment]
@MSeifert04 it's a newline at the end of the string: the `\n` in the above example

**Comment by keflavich** [comment]
This problem has gotten a bit worse for me.  I tried the obvious workaround of editing the HDU inplace:

```
    fh = fits.open(fn)
    fh[0].header['ORIGIN'] = 'CASA 4.7.0-REL (r38335)'
    print('Header origin: "{0}"'.format(fh[0].header['ORIGIN'])) # confirms no newline
    wcs.WCS(fh[0].header) #crashes!!
```

but it results in the same traceback:
```
Header origin: "CASA 4.7.0-REL (r38335)"
Traceback (most recent call last):
  File "<ipython-input-72-2c3f12ce28f0>", line 16, in <module>
    wcs.WCS(fh[0].header)
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/wcs/wcs.py", line 391, in __init__
    try:
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/header.py", line 639, in tostring
    s += sep + _pad('END')
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/card.py", line 461, in __str__
    return self.image
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/card.py", line 765, in image
    self.verify('fix')
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/verify.py", line 74, in verify
    errs = self._verify(opt)
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/card.py", line 1392, in _verify
    fix=self._fix_value))
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/verify.py", line 48, in run_option
    fix()
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/card.py", line 1131, in _fix_value
    self.value = valuecomment
  File "/users/aginsbur/anaconda/lib/python3.5/site-packages/astropy/io/fits/card.py", line 582, in value
    'ASCII or non-printable characters.' % value)
ValueError: FITS header values must contain standard printable ASCII characters; "'CASA 4.7.0-REL (r38335)\n'" contains characters not representable in ASCII or non-printable characters.
```

How is WCS accessing some older version of the header?  It's as if the header isn't being updated.

**Comment by pllim** [comment]
Try `mode=update` or... make a copy of that header, modify the copy, and pass in the modified copy to WCS. Does either of those work?

**Comment by keflavich** [comment]
@pllim Thanks for the ideas.  Copy fails (same).  It actually shows that WCS is not accessing the same header... which I don't understand at all.  

```
    fh = fits.open(fn)
    fh[0].header['ORIGIN'] = 'CASA 4.7.0-REL (r38335)'
    print('Header origin: "{0}"'.format(fh[0].header['ORIGIN']))
    header = copy.copy(fh[0].header)
    header['ORIGIN'] = 'BLANK'

ValueError: FITS header values must contain standard printable ASCII characters; "'CASA 4.7.0-REL (r38335)\n'" contains characters not representable in ASCII or non-printable characters.
```

same problem with `fh=fits.open(fn, mode='update')`

**Comment by keflavich** [comment]
Stranger still... in my first traceback, the failure is on line 391 of `wcs.py`, which implies the header I'm passing is not a `fits.Header` instance, even though it clearly is (which I can even verify using `%debug`).

**Comment by keflavich** [comment]
This has some heisenbug elements to it.  If I run this as a block of code (i.e., with `%cpaste`), I get the error.  If I run each command in turn (just paste the whole thing on the command line), it doesn't happen.  Maybe it's something peculiar ipython is doing.  The overall issue remains important, though.

**Comment by mcara** [comment]
@keflavich Is there a way to get your test image?

**Comment by mcara** [comment]
@keflavich I have tried to reproduce your example from https://github.com/astropy/astropy/issues/5396#issuecomment-252284805 (using an STScI image modified to contain ``\n`` in the ``ORIGIN`` string value) and I cannot reproduce your crash: WCS can be successfully constructed from a "corrected" header. Are you sure you are not attempting to create the WCS by providing the file name instead of the header to the WCS constructor? It looks to me that you are reloading the old file...

**Comment by keflavich** [comment]
@mcara - in your example, you get a traceback the first time you access the header, but not the second.  I get the same, even if I don't correct the header.  i.e.:

```
fh = fits.open(fn)
fh[0].header.tostring() # produces exception
fh[0].header.tostring() # does not produce exception
```

Can you reproduce that behavior?

**Comment by mcara** [comment]
@keflavich Yes, I can reproduce this behavior.

**Comment by mcara** [comment]
If in my previous example I do not attempt to create a WCS immediately after opening FITS file, then attempting to create a WCS after correcting the header still crashes (as you have reported):
```python
>>> h = fits.open('test1n.fits')
>>> h[0].header['origin'] = 'goodval'
>>> w = wcs.WCS(h[0].header)  # <- crashes
```
This is exciting! 😄

**Comment by mcara** [comment]
... On the other hand, if I use ``tostring()`` instead of attempting to create a WCS, I get the following:
```python
>>> from astropy.io import fits
>>> from astropy import wcs
>>> import numpy as np
>>> h = fits.open('test1n.fits')
>>> h[0].header['origin'] = 'goodval'
>>> h[0].header.tostring()
WARNING: VerifyWarning: Verification reported errors: [astropy.io.fits.verify]
WARNING: VerifyWarning: Card 'ORIGIN' is not FITS standard (invalid value string: "'HSTIO/CFITSIO March 2010\n'").  Fixed 'ORIGIN' card to meet the FITS standard. [astropy.io.fits.verify]
WARNING: VerifyWarning: Note: astropy.io.fits uses zero-based indexing.
 [astropy.io.fits.verify]
"SIMPLE  =                    T / conforms to FITS standard                      BITPIX  =                  -32 / array data type                                NAXIS   =                    2 / number of array dimensions                     NAXIS1  =                 2061                                                  NAXIS2  =                 1921                                                  COMMENT   FITS (Flexible Image Transport System) format is defined in 'AstronomyCOMMENT   and Astrophysics', volume 376, page 359; bibcode: 2001A&A...376..359H ORIGIN  = '''HSTIO '           / CFITSIO March 2010\n'                           DATE    = '2014-09
```
(scroll to the right to see ``'ORIGIN'``). So, it seems that ``h[0].header['origin'] = 'goodval'`` had no effect!

**Comment by keflavich** [comment]
Interesting!  If I do the same, I get a crash on `tostring` the first time I call it, but the second time it shows that the keyword was successfully changed:

```
>>> fh = fits.open(fn)
>>> fh[0].header['ORIGIN']='goodval'
# exception:
# ValueError: FITS header values must contain standard printable ASCII characters; "'CASA 4.7.0-REL (r38335)\n'" contains characters not representable in ASCII or non-printable characters.
>>> fh[0].header.tostring()
# ... snip ... ORIGIN  = \'goodval \' ... snip ...
```

**Comment by mcara** [comment]
Well, I think this has to be a bug in ``fits.io`` but it would probably be better to open a new issue as this is not logically connected to CASA's bug.

**Comment by keflavich** [comment]
Hmm, right.  There are two independent issues here.  *This* issue is about how best to use files that have illegal header keywords.

**Comment by keflavich** [comment]
The upstream fix has been implemented, so most likely we will not hear much about this specific error - we caught it early enough that probably no data have been distributed with the bad headers.  However, a preferred workaround would still be helpful to have here.

**Comment by mcara** [comment]
I am still worried that one cannot change header values right away - it seems that header is re-loading old values from somewhere...

**Comment by keflavich** [comment]
@mcara: agreed.   I opened the new issue here now: https://github.com/astropy/astropy/issues/5408

**Comment by bsipocz** [comment]
This seems to be fixed by #6084 based on https://github.com/astropy/astropy/pull/6084#issuecomment-302334340. 

Please feel free to reopen if I misunderstood and it didn't get fixed.
